### PR TITLE
Fix #35272: Don't replace typedefs in extern c scope

### LINKED ISSUE #35272: modernize-use-using: don't change code in extern "C" {}
|  |  |
| --- | --- |
| Bugzilla Link | [35924](https://llvm.org/bz35924) |
| Version | unspecified |
| OS | Linux |
| Reporter | LLVM Bugzilla Contributor |
| CC | @dkalowsk,@JonasToth,@EugeneZelenko,@gburgessiv,@hokein,@johnmcfarlane,@LegalizeAdulthood,@Ralender,@tinti |

## Extended Description 
clang-tidy generates warnings specific to C++ for code that lies within the scope of an extern "C" {}

E.g. consider the following code

/* example.cpp */

/* This block simulates a C header */
#ifdef __cplusplus
extern "C" {
#endif

typedef int my_int_t;

#ifdef __cplusplus
} // extern "C"
#endif

// This block simulates a C++ implementation/use of the C header's API
#include <iostream>

int main()
{
  my_int_t value{42};
  std::cout << value << std::endl;
  return 0;
}

Compile with:

$ clang++ --std=c++14 -Wall example.cpp -o example

Run clang-tidy as:

$ clang-tidy example.cpp -checks='*' -- -std=c++14 -Wall

gives:

5307 warnings generated.
/home/duncan/temp/example.cpp:5:1: warning: use 'using' instead of 'typedef' [modernize-use-using]
typedef int my_int_t;
^~~~~~~~~~~~~~~~~~~~~
using my_int_t = int
Suppressed 5306 warnings (5306 in non-user code).
Use -header-filter=.* to display errors from all non-system headers. Use -system-headers to display errors from system headers as well.

I don't think C++ checks should be applied to code within the scope of an extern "C" block.

### PR DESCRIPTION
**Pull request by da-viper**
Fixes #35272

### CODE REVIEW

**Review by HerrCai0907 — COMMENTED** [review]
please update release note and doc

**Comment by PiotrZSL** [comment]
I got mixed feelings about this. Simply we can use 'using' within extern "C", there is nothing wrong with this. using/typedef doesn't haven linkage, that just type alias. And extern "C" does not impact it. In oryginal issue I think someone mistook extern "C" into something where only "C" code should be put, and that's not true. All what extern "C" does it's using C style symbol mangling for functions inside.

The only reason to not use 'using' in extern 'C' would be just readability, to avoid some confusion.
As for this change, release notes & doc is needed.

**Comment by PiotrZSL** [comment]
> 2. where do I find the release notes.
clang-tools-extra/docs/ReleaseNotes.rst

File: clang-tools-extra/docs/ReleaseNotes.rst [hunk 1]
```diff
@@ -285,6 +285,10 @@ Changes in existing checks
   <clang-tidy/checks/modernize/use-using>` check to fix function pointer and
   forward declared ``typedef`` correctly.
 
+- Improved :doc:`modernize-use-using
```
Comment (PiotrZSL) [inline clang-tools-extra/docs/ReleaseNotes.rst:288]: merge with previous entry for this check.

File: clang-tools-extra/docs/ReleaseNotes.rst [hunk 2]
```diff
@@ -325,6 +329,7 @@ Changes in existing checks
   identify calls to static member functions with out-of-class inline definitions.
 
 
+
```
Comment (PiotrZSL) [inline clang-tools-extra/docs/ReleaseNotes.rst:332]: remove added empty line

File: clang-tools-extra/docs/clang-tidy/checks/modernize/use-using.rst [hunk 3]
```diff
@@ -28,6 +28,15 @@ After:
   using R_t = struct { int a; };
   using R_p = R_t*;
 
+The checker ignores `typedef` within `extern "C" { ... }` blocks.
+
+.. code-block:: c++
+
+  extern "C" {
+
```
Comment (PiotrZSL) [inline clang-tools-extra/docs/clang-tidy/checks/modernize/use-using.rst:36]: remove empty line

File: clang-tools-extra/docs/clang-tidy/checks/modernize/use-using.rst [hunk 4]
```diff
@@ -28,6 +28,15 @@ After:
   using R_t = struct { int a; };
   using R_p = R_t*;
 
+The checker ignores `typedef` within `extern "C" { ... }` blocks.
```
Comment (PiotrZSL) [inline clang-tools-extra/docs/clang-tidy/checks/modernize/use-using.rst:31]: Maybe this should be put into configuration option ? To give an option to preserve current behavior.

Comment (Da-Viper) [reply clang-tools-extra/docs/clang-tidy/checks/modernize/use-using.rst:31]: Would it be preferred to make this an option ?

File: clang-tools-extra/docs/clang-tidy/checks/modernize/use-using.rst (hunk 3 above)
Comment (Da-Viper) [reply clang-tools-extra/docs/clang-tidy/checks/modernize/use-using.rst]: done

File: clang-tools-extra/docs/ReleaseNotes.rst (hunk 1 above)
Comment (Da-Viper) [reply clang-tools-extra/docs/ReleaseNotes.rst]: done

File: clang-tools-extra/docs/clang-tidy/checks/modernize/use-using.rst (hunk 4 above)
Comment (PiotrZSL) [reply clang-tools-extra/docs/clang-tidy/checks/modernize/use-using.rst:31]: I think so, for example I would like to enable this for my projects.
Ignoring extern "C" makes only sense if same header is compile in C and C++ mode.
If header is compiled only in C++ mode, but uses extern "C" simply to properly handle symbols mangling, then there is no need to ignore such typedefs.

File: clang-tools-extra/docs/clang-tidy/checks/modernize/use-using.rst [hunk 5]
```diff
@@ -28,6 +28,13 @@ After:
...
+
+.. code-block:: c++
+  extern "C" {
```
Comment (PiotrZSL) [inline clang-tools-extra/docs/clang-tidy/checks/modernize/use-using.rst:34]: put empty line here, this is why documentation fails

File: clang-tools-extra/docs/ReleaseNotes.rst [hunk 6]
```diff
@@ -318,7 +318,8 @@ Changes in existing checks
 
 - Improved :doc:`modernize-use-using
   <clang-tidy/checks/modernize/use-using>` check to fix function pointer and
-  forward declared ``typedef`` correctly.
+  forward declared ``typedef`` correctly. Ignore ``typedef`` declaration in
```
Comment (PiotrZSL) [inline clang-tools-extra/docs/ReleaseNotes.rst:321]: Actually release notes should say that new option IgnoreExternC were added to control whatever typedefs in extern C should be ignored.

**Review by PiotrZSL — APPROVED** [review]
Loooks, good, minor nit in release notes.

**Pull request merged**